File: scripts/build-search-index.ts
/**
 * build-search-index.ts
 *
 * public/data/books-{l1id}.json から検索用インデックスを生成し
 * public/data/search-index.json に書き出す。
 *
 * ※ build-split-index.ts の後に実行すること（prebuild順序依存）
 */

import { writeFileSync, readFileSync } from "fs";
import { join } from "path";
import { CATEGORY_TREE } from "../src/lib/categories";

// ── 型定義 ───────────────────────────────────────────────────────

interface SplitBook {
  id: string;
  title: string;
  authors: string[];
  publisher?: string;
  publishedDate?: string;
  isbn13?: string;
  keywords: string[];
  pageCount?: number;
  estimatedReadingHours?: number;
  thumbnailUrl?: string;
  sourceIds?: { googleBooksId?: string };
}

export interface SearchEntry {
  id: string;
  title: string;
  authors: string[];
  publisher?: string;
  publishedDate?: string;   // YYYY-MM 形式
  isbn13?: string;
  googleBooksId?: string;
  thumbnailUrl?: string;
  pageCount?: number;
  estimatedReadingHours?: number;
  keywords: string[];
  l1Id: string;
}

// ── メイン処理 ────────────────────────────────────────────────────

const dataDir = join(process.cwd(), "public", "data");
const allEntries: SearchEntry[] = [];

for (const l1 of CATEGORY_TREE) {
  const filePath = join(dataDir, `books-${l1.id}.json`);
  let books: SplitBook[];
  try {
    books = JSON.parse(readFileSync(filePath, "utf-8"));
  } catch {
    continue;
  }

  for (const b of books) {
    if (!b.title || !b.authors?.length) continue;

    const entry: SearchEntry = {
      id: b.id,
      title: b.title,
      authors: b.authors,
      keywords: b.keywords.slice(0, 10),
      l1Id: l1.id,
    };

    if (b.publisher) entry.publisher = b.publisher;
    if (b.publishedDate) entry.publishedDate = b.publishedDate.slice(0, 7);
    if (b.isbn13) entry.isbn13 = b.isbn13;
    if (b.sourceIds?.googleBooksId) entry.googleBooksId = b.sourceIds.googleBooksId;
    if (b.thumbnailUrl) entry.thumbnailUrl = b.thumbnailUrl;
    if (b.pageCount) entry.pageCount = b.pageCount;
    if (b.estimatedReadingHours) {
      entry.estimatedReadingHours = parseFloat(b.estimatedReadingHours.toFixed(1));
    }

    allEntries.push(entry);
  }
}

writeFileSync(
  join(dataDir, "search-index.json"),
  JSON.stringify(allEntries),
);

console.log(`✓ search-index.json  (${allEntries.length}冊)`);
console.log("Done.");

File: scripts/build-search-index.py
#!/usr/bin/env python3
"""
build-search-index.ts の Python版。
public/data/books-{l1id}.json から検索用インデックスを生成。
"""

import json
import os

ROOT = os.path.dirname(os.path.dirname(os.path.abspath(__file__)))
DATA_DIR = os.path.join(ROOT, "public", "data")

L1_IDS = ["manga", "novel", "business", "self-help", "tech", "science",
           "philosophy", "psychology", "history", "investing"]

all_entries = []

for l1_id in L1_IDS:
    fpath = os.path.join(DATA_DIR, f"books-{l1_id}.json")
    if not os.path.exists(fpath):
        continue

    with open(fpath, encoding="utf-8") as f:
        books = json.load(f)

    for b in books:
        if not b.get("title") or not b.get("authors"):
            continue

        entry = {
            "id": b["id"],
            "title": b["title"],
            "authors": b["authors"],
            "keywords": b.get("keywords", [])[:10],
            "l1Id": l1_id,
        }

        if b.get("publisher"): entry["publisher"] = b["publisher"]
        if b.get("publishedDate"): entry["publishedDate"] = b["publishedDate"][:7]
        if b.get("isbn13"): entry["isbn13"] = b["isbn13"]
        if b.get("sourceIds", {}).get("googleBooksId"):
            entry["googleBooksId"] = b["sourceIds"]["googleBooksId"]
        if b.get("thumbnailUrl"): entry["thumbnailUrl"] = b["thumbnailUrl"]
        if b.get("pageCount"): entry["pageCount"] = b["pageCount"]
        if b.get("estimatedReadingHours"):
            entry["estimatedReadingHours"] = round(b["estimatedReadingHours"], 1)

        all_entries.append(entry)

with open(os.path.join(DATA_DIR, "search-index.json"), "w", encoding="utf-8") as f:
    json.dump(all_entries, f, ensure_ascii=False, separators=(",", ":"))

print(f"✓ search-index.json  ({len(all_entries)}冊)")
